pkg/roachprod/vm/aws/aws.go

```diff
@@ -278,8 +278,57 @@ func (p *Provider) GetHostErrorVMs(
 	return nil, nil
 }
 
-func (p *Provider) GetVMSpecs(vms vm.List) ([]map[string]interface{}, error) {
-	return nil, nil
+func (p *Provider) GetVMSpecs(
+	l *logger.Logger, vms vm.List,
+) (map[string]map[string]interface{}, error) {
+	if vms == nil {
+		return nil, errors.New("vms cannot be nil")
+	}
+
+	byRegion, err := regionMap(vms)
+	if err != nil {
+		return nil, err
+	}
+
+	var instanceList []map[string]interface{}
+	// Extract the spec of all VMs and create a map from VM InstanceId to spec.
+	vmSpecs := make(map[string]map[string]interface{})
+	for region, list := range byRegion {
+		args := []string{
+			"ec2", "describe-instances",
+			"--region", region,
+			"--query", "Reservations[].Instances[]",
+			"--instance-ids",
+		}
+		args = append(args, list.ProviderIDs()...)
+		if err := p.runJSONCommand(l, args, &instanceList); err != nil {
+			return nil, errors.Wrapf(err, "error describing instances in region %s: ", region)
+		}
+
+		if len(instanceList) == 0 {
+			l.Errorf("no instances found in region %s", region)
+			continue
+		}
+
+		for _, instance := range instanceList {
+			instanceName := ""
+			for _, tag := range instance["Tags"].([]interface{}) {
```

Comment:
Hmm, I don't think this would help. `MakeMap` already has a `Tags` type while I have `[]interface {}`. I do something similar anyway, where I loop over each `interface {}` and convert it into `map[string]interface{}`. However, I can probably convert each to `map[string]string` like in the function though 🤔.

Reply:
It might be a bit cleaner to use `DescribeInstancesOutput`; see `listRegion`. Roughly, it would look something like this,

```
vmSpecs := make(map[string]map[string]interface{})
	for region, list := range byRegion {
		args := []string{
			"ec2", "describe-instances",
			"--region", region,
			"--instance-ids",
		}
		args = append(args, list.ProviderIDs()...)
		var describeInstancesResponse DescribeInstancesOutput
		err := p.runJSONCommand(l, args, &describeInstancesResponse)
		if err != nil {
			return nil, err
		}
		for _, r := range describeInstancesResponse.Reservations {
			for _, inst := range r.Instances {
				nameTag := slices.DeleteFunc(inst.Tags, func(tag Tag) bool {
					return tag.Key != "Name"
				})
				if len(nameTag) == 1 {
					inrec, _ := json.Marshal(inst)
					var m map[string]interface{}
					json.Unmarshal(inrec, &m)
					vmSpecs[nameTag[0].Value] = m
				}
			}
		}
	}
	return vmSpecs, nil
```

Specifically, I left out proper error checking and further refactoring :)

Reply:
Definitely cleaner. Updated with error checks.